File: fixtures/sample-android-firebase-app/app/src/main/java/com/example/myapp/auth/AuthManager.kt
package com.example.myapp.auth

import android.util.Log
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.tasks.await

/**
 * Manages Firebase authentication state for the app
 */
class AuthManager {
    private val auth: FirebaseAuth = FirebaseAuth.getInstance()

    private val _currentUser = MutableStateFlow<FirebaseUser?>(null)
    val currentUser: StateFlow<FirebaseUser?> = _currentUser.asStateFlow()

    private val _isAuthenticated = MutableStateFlow(false)
    val isAuthenticated: StateFlow<Boolean> = _isAuthenticated.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    init {
        // Set up auth state listener
        auth.addAuthStateListener { firebaseAuth ->
            _currentUser.value = firebaseAuth.currentUser
            _isAuthenticated.value = firebaseAuth.currentUser != null
        }
    }

    /**
     * Sign in with email and password
     */
    suspend fun signIn(email: String, password: String): Result<FirebaseUser> {
        return try {
            _isLoading.value = true
            _errorMessage.value = null

            val result = auth.signInWithEmailAndPassword(email, password).await()
            val user = result.user ?: throw Exception("User is null after sign in")

            _currentUser.value = user
            _isAuthenticated.value = true
            _isLoading.value = false

            Result.success(user)
        } catch (e: Exception) {
            Log.e(TAG, "Sign in failed", e)
            _errorMessage.value = e.localizedMessage
            _isLoading.value = false
            Result.failure(e)
        }
    }

    /**
     * Create a new account with email and password
     */
    suspend fun createAccount(email: String, password: String): Result<FirebaseUser> {
        return try {
            _isLoading.value = true
            _errorMessage.value = null

            val result = auth.createUserWithEmailAndPassword(email, password).await()
            val user = result.user ?: throw Exception("User is null after account creation")

            _currentUser.value = user
            _isAuthenticated.value = true
            _isLoading.value = false

            Result.success(user)
        } catch (e: Exception) {
            Log.e(TAG, "Account creation failed", e)
            _errorMessage.value = e.localizedMessage
            _isLoading.value = false
            Result.failure(e)
        }
    }

    /**
     * Sign out the current user
     */
    fun signOut() {
        auth.signOut()
        _currentUser.value = null
        _isAuthenticated.value = false
    }

    /**
     * Send password reset email
     */
    suspend fun sendPasswordReset(email: String): Result<Unit> {
        return try {
            _isLoading.value = true
            _errorMessage.value = null

            auth.sendPasswordResetEmail(email).await()
            _isLoading.value = false

            Result.success(Unit)
        } catch (e: Exception) {
            Log.e(TAG, "Password reset failed", e)
            _errorMessage.value = e.localizedMessage
            _isLoading.value = false
            Result.failure(e)
        }
    }

    companion object {
        private const val TAG = "AuthManager"

        @Volatile
        private var instance: AuthManager? = null

        fun getInstance(): AuthManager {
            return instance ?: synchronized(this) {
                instance ?: AuthManager().also { instance = it }
            }
        }
    }
}

File: fixtures/sample-ios-firebase-app/MyApp/Auth/AuthManager.swift
import Foundation
import FirebaseAuth
import FirebaseCore

/// Manages authentication state for the app
class AuthManager: ObservableObject {
    @Published var user: User?
    @Published var isAuthenticated: Bool = false
    @Published var isLoading: Bool = false
    @Published var errorMessage: String?

    private var authStateHandler: AuthStateDidChangeListenerHandle?

    static let shared = AuthManager()

    init() {
        setupAuthStateListener()
    }

    deinit {
        if let handle = authStateHandler {
            Auth.auth().removeStateDidChangeListener(handle)
        }
    }

    // MARK: - Auth State Listener

    private func setupAuthStateListener() {
        authStateHandler = Auth.auth().addStateDidChangeListener { [weak self] _, user in
            DispatchQueue.main.async {
                self?.user = user
                self?.isAuthenticated = user != nil
            }
        }
    }

    // MARK: - Sign In with Email

    func signIn(email: String, password: String) async throws {
        isLoading = true
        errorMessage = nil

        do {
            let result = try await Auth.auth().signIn(withEmail: email, password: password)
            DispatchQueue.main.async {
                self.user = result.user
                self.isAuthenticated = true
                self.isLoading = false
            }
        } catch {
            DispatchQueue.main.async {
                self.errorMessage = error.localizedDescription
                self.isLoading = false
            }
            throw error
        }
    }

    // MARK: - Create Account

    func createAccount(email: String, password: String) async throws {
        isLoading = true
        errorMessage = nil

        do {
            let result = try await Auth.auth().createUser(withEmail: email, password: password)
            DispatchQueue.main.async {
                self.user = result.user
                self.isAuthenticated = true
                self.isLoading = false
            }
        } catch {
            DispatchQueue.main.async {
                self.errorMessage = error.localizedDescription
                self.isLoading = false
            }
            throw error
        }
    }

    // MARK: - Sign Out

    func signOut() throws {
        try Auth.auth().signOut()
        DispatchQueue.main.async {
            self.user = nil
            self.isAuthenticated = false
        }
    }

    // MARK: - Password Reset

    func sendPasswordReset(email: String) async throws {
        try await Auth.auth().sendPasswordReset(withEmail: email)
    }
}
